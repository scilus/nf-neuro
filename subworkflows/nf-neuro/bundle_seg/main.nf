import org.apache.commons.io.FileUtils
import java.nio.file.Files
import java.nio.file.Paths
import java.util.zip.ZipFile

include { REGISTRATION_ANTS                 } from '../../../modules/nf-neuro/registration/ants/main'
include { BUNDLE_RECOGNIZE                  } from '../../../modules/nf-neuro/bundle/recognize/main'

def fetch_bundleseg_atlas(atlasUrl, configUrl, dest) {

    def atlas = new File("$dest/atlas.zip").withOutputStream { out ->
        new URL(atlasUrl).withInputStream { from -> out << from; }
    }

    def config = new File("$dest/config.zip").withOutputStream { out ->
        new URL(configUrl).withInputStream { from -> out << from; }
    }

    def atlasFile = new ZipFile("$dest/atlas.zip")
    atlasFile.entries().each { it ->
        def path = Paths.get("$dest/atlas/" + it.name)
        if(it.directory){
            Files.createDirectories(path)
        }
        else {
            def parentDir = path.getParent()
            if (!Files.exists(parentDir)) {
                Files.createDirectories(parentDir)
            }
            Files.copy(atlasFile.getInputStream(it), path)
        }
    }

    def configFile = new ZipFile("$dest/config.zip")
    configFile.entries().each { it ->
        def path = Paths.get("$dest/config/" + it.name)
        if(it.directory){
            Files.createDirectories(path)
        }
        else {
            def parentDir = path.getParent()
            if (!Files.exists(parentDir)) {
                Files.createDirectories(parentDir)
            }
            Files.copy(configFile.getInputStream(it), path)
        }
    }
}

workflow BUNDLE_SEG {

    take:
        ch_fa               // channel: [ val(meta), [ fa ] ]
        ch_tractogram       // channel: [ val(meta), [ tractogram ] ]

    main:

        ch_versions = Channel.empty()

        // ** Setting up Atlas reference channels. ** //
        if ( params.atlas_directory ) {
            atlas_anat = Channel.fromPath("$atlas_directory/atlas/mni_masked.nii.gz")
            atlas_config = Channel.fromPath("$atlas_directory/config/config_fss_1.json")
            atlas_average = Channel.fromPath("$atlas_directory/atlas/atlas/")
        }
        else {
            fetch_bundleseg_atlas(  "https://zenodo.org/records/10103446/files/atlas.zip?download=1",
                                    "https://zenodo.org/records/10103446/files/config.zip?download=1",
                                    "${workflow.workDir}/")
            atlas_anat = Channel.fromPath("$workflow.workDir/atlas/mni_masked.nii.gz")
            atlas_config = Channel.fromPath("$workflow.workDir/config/config_fss_1.json")
            atlas_average = Channel.fromPath("$workflow.workDir/atlas/atlas/")
        }

        // ** Register the atlas to subject's space. Set up atlas file as moving image ** //
        // ** and subject anat as fixed image.                                         ** //
        ch_register =  ch_fa.combine(atlas_anat)
                            .map{ it + [[]] }

        REGISTRATION_ANTS ( ch_register )
        ch_versions = ch_versions.mix(REGISTRATION_ANTS.out.versions.first())

        // ** Perform bundle recognition and segmentation ** //
        ch_recognize_bundle =  ch_tractogram.join(REGISTRATION_ANTS.out.transfo_image)
                                            .combine(atlas_config)
                                            .combine(atlas_average)

        BUNDLE_RECOGNIZE ( ch_recognize_bundle )
        ch_versions = ch_versions.mix(BUNDLE_RECOGNIZE.out.versions.first())


    emit:
        bundles = BUNDLE_RECOGNIZE.out.bundles              // channel: [ val(meta), [ bundles ] ]

        versions = ch_versions                              // channel: [ versions.yml ]
}
