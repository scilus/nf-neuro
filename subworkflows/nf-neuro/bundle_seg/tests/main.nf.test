nextflow_workflow {

    name "Test Subworkflow BUNDLE_SEG"
    script "../main.nf"
    workflow "BUNDLE_SEG"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/bundle_seg"

    tag "bundle/recognize"
    tag "registration/ants"

    tag "load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "tracking.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("rbx - base - no_atlas_folder") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/fa.nii.gz")
                    ]
                }
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ],
                        file("\${test_data_directory}/pft.trk")
                    ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
