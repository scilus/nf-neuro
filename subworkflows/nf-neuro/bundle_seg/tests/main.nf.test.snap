{
    "rbx - base - no_atlas_folder": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__AF_L_cleaned.trk:md5,b9a6567efffa206202f37c11afea2391",
                            "test__CC_Fr_2_cleaned.trk:md5,ad18f4415d6d5baae1ec81359ccffecb",
                            "test__CC_Pr_Po_cleaned.trk:md5,dca763cddc2decfa64baf452cc035b5a",
                            "test__CG_R_An_cleaned.trk:md5,79c5f2469b1fcae2599247ced8d68e07",
                            "test__CG_R_cleaned.trk:md5,79c5f2469b1fcae2599247ced8d68e07",
                            "test__FPT_L_Brainstem_cleaned.trk:md5,d6c9af62d9ad65d95257093148bf4d9a",
                            "test__FPT_L_cleaned.trk:md5,d6c9af62d9ad65d95257093148bf4d9a",
                            "test__ILF_L_cleaned.trk:md5,00f9f6e33e59bfe8bbe3a5afd18b0f2e",
                            "test__MCP_cleaned.trk:md5,ed86370180b35e125ae410aa3d520625",
                            "test__OR_ML_L_cleaned.trk:md5,721b658ed8a4de104c4e7c207d4fe170",
                            "test__POPT_R_Brainstem_cleaned.trk:md5,6c07d476650f4bca894c4c2454a1783f",
                            "test__POPT_R_cleaned.trk:md5,6c07d476650f4bca894c4c2454a1783f",
                            "test__PYT_L_cleaned.trk:md5,ac9973ef987c34ecee0328004033ea46",
                            "test__PYT_R_Brainstem_cleaned.trk:md5,28b4be34422e356030b6b081efca31d5",
                            "test__PYT_R_cleaned.trk:md5,50f8127bbaab6f4b9fea790a7c11679d",
                            "test__SLF_L_cleaned.trk:md5,3660a62ee7e0e413f7417e80c1842e60"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,b5ff3e60063810aea9f911b2a439b592",
                    "versions.yml:md5,e63f6133bb2e78ce211054111993be59"
                ],
                "bundles": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        [
                            "test__AF_L_cleaned.trk:md5,b9a6567efffa206202f37c11afea2391",
                            "test__CC_Fr_2_cleaned.trk:md5,ad18f4415d6d5baae1ec81359ccffecb",
                            "test__CC_Pr_Po_cleaned.trk:md5,dca763cddc2decfa64baf452cc035b5a",
                            "test__CG_R_An_cleaned.trk:md5,79c5f2469b1fcae2599247ced8d68e07",
                            "test__CG_R_cleaned.trk:md5,79c5f2469b1fcae2599247ced8d68e07",
                            "test__FPT_L_Brainstem_cleaned.trk:md5,d6c9af62d9ad65d95257093148bf4d9a",
                            "test__FPT_L_cleaned.trk:md5,d6c9af62d9ad65d95257093148bf4d9a",
                            "test__ILF_L_cleaned.trk:md5,00f9f6e33e59bfe8bbe3a5afd18b0f2e",
                            "test__MCP_cleaned.trk:md5,ed86370180b35e125ae410aa3d520625",
                            "test__OR_ML_L_cleaned.trk:md5,721b658ed8a4de104c4e7c207d4fe170",
                            "test__POPT_R_Brainstem_cleaned.trk:md5,6c07d476650f4bca894c4c2454a1783f",
                            "test__POPT_R_cleaned.trk:md5,6c07d476650f4bca894c4c2454a1783f",
                            "test__PYT_L_cleaned.trk:md5,ac9973ef987c34ecee0328004033ea46",
                            "test__PYT_R_Brainstem_cleaned.trk:md5,28b4be34422e356030b6b081efca31d5",
                            "test__PYT_R_cleaned.trk:md5,50f8127bbaab6f4b9fea790a7c11679d",
                            "test__SLF_L_cleaned.trk:md5,3660a62ee7e0e413f7417e80c1842e60"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b5ff3e60063810aea9f911b2a439b592",
                    "versions.yml:md5,e63f6133bb2e78ce211054111993be59"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-24T20:26:57.178201"
    }
}